js/index.js

```diff
@@ -230,71 +254,59 @@
       eventObj.trigger('update-url', {'schema' : '' , 'record' : ''});
       eventObj.trigger('reset-layout');
       window.type = undefined;
-      $('#random').addClass('-disabled-');
       return false;
     });
 
     $('#uploadLink').click(function(e) {
-      console.log("uploading");
       e.preventDefault();
-      $("#upload").trigger('click');
-      return false;
+      uploadElement.trigger('click');
+      return false; // So that it doesn't show the content of the file.
     });
 
-    $('#upload').on("change", function(e) {
-      console.log("new file uploading");
-      var files = $('#upload')[0].files;
-      if(!!files && files.length > 0) {
-        var file = files[0]; //TODO: make sure only one file can be selected.
-        var reader = new FileReader();
-        reader.readAsText(file, "UTF-8");
-        reader.onload = function (evt) {
-
-          eventObj.trigger('schema-loaded', evt.target.result);
-          console.log(evt.target.result);
-        }
-        reader.onerror = function (evt) {
-          console.log("error reading file.");
+    uploadElement.on("change", function(e) {
+      var files = $(uploadElement)[0].files;
+      eventObj.trigger('schema-uploaded', files);
+    });
+
+    $(document).click(function(e) {
+      if(!$(e.target).closest('#schema').length) {
+        if (!schemaElement.text()){
+          template.show();
+          hideError(schemaErrorElement);
         }
       }
-
     });
 
     function populateFromQuery() {
-      var s = readQueryValue(location.href, 'schema');
-      if(!!s) {
-        s = decodeURIComponent(s);
-        $(schemaElement).text(s);
-        eventObj.trigger('schema-changed');
+      var s = urlUtils.readValue('schema');
+      if(s) {
+        eventObj.trigger('schema-changed', JSON.parse(s));
       }
       
-      var record = readQueryValue(location.href, 'record');
-      if(!!record) {
-        record = decodeURIComponent(record);
+      var record = urlUtils.readValue('record');
+      if(record) {
         decode(record);
         setOutputText(record);
       }
-
-      if (!!s && !record) {
-        generateRandom();
-      }
     }
     
     /**
     * Will save cursor position inside element `elemId` before running callback function f,
     * and restores it after f is finished. 
     */ 
-    function runPreservingCursorPosition(elementId, f) {
+    function runPreservingCursorPosition(elementId, f, options) {
+      var context = options ? options.context: undefined;
```

Comment:
Unless you need the context to be exactly `undefined`, this is often rewritten as `options && options.context`. (Same below.)